### PR TITLE
Split network into several elements

### PR DESCRIPTION
**Pull request by archendel**

### CODE REVIEW

File: src/components/network/use-equipments.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export const useEquipments = (equipmentType) => {
+    const dispatch = useDispatch();
+    const equipments = useSelector(
+        (state) => state.spreadsheetNetwork[equipmentType.resource]
+    );
+    const studyUuid = useSelector((state) => state.studyUuid);
+    const currentNode = useSelector((state) => state.currentTreeNode);
+
+    const isFetching = equipments == null;
```
Comment (flomillot) [inline src/components/network/use-equipments.js:13]:
```suggestion
    const isFetching = !equipments;
```

Comment (flomillot) [reply src/components/network/use-equipments.js:13]: Should be the same but for me easier to read

File: src/components/network/use-equipments.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    useEffect(() => {
+        if (equipments == null) {
```
Comment (flomillot) [inline src/components/network/use-equipments.js:16]: Same as above ?

File: src/components/network/use-equipments.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            Promise.all(
+                equipmentType.type.fetchers.map((fetcher) =>
+                    fetcher(studyUuid, currentNode.id)
+                )
+            ).then((results) => {
+                const equipments = [].concat(...results);
```
Comment (flomillot) [inline src/components/network/use-equipments.js:22]: Why this line ? `results` is already an array no ?

Comment (flomillot) [reply src/components/network/use-equipments.js:22]: After discussion, maybe use `.flat` instead

Comment (archendel) [reply src/components/network/use-equipments.js:22]: fixed

File: src/components/spreadsheet/table-wrapper.js [hunk 4]
```diff
@@ -297,25 +293,24 @@ const TableWrapper = (props) => {
...
-        },
-        [props.disabled, props.network]
-    );
+        const datasourceRows = equipments;
+
+        if (!datasourceRows) {
+            return [];
+        }
+
+        //the method returns a new array so that the table component detects its data changed thus rerendering its rows
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/components/spreadsheet/table-wrapper.js:311]: done

File: src/components/spreadsheet/table-wrapper.js [hunk 5]
```diff
@@ -297,25 +293,24 @@ const TableWrapper = (props) => {
...
+        if (props.disabled) {
+            return [];
+        }
 
-            //the method returns a new array so that the table component detects its data changed thus rerendering its rows
-            return [...datasourceRows];
-        },
-        [props.disabled, props.network]
-    );
+        const datasourceRows = equipments;
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/components/spreadsheet/table-wrapper.js:305]: fixed

File: src/components/spreadsheet/table-wrapper.js [hunk 6]
```diff
@@ -297,25 +293,24 @@ const TableWrapper = (props) => {
...
-                : props.network[tableDefinition.resource];
+    const { equipments, isFetching } = useEquipments(
+        TABLES_DEFINITION_INDEXES.get(tabIndex)
+    );
 
-            if (!datasourceRows) {
-                return [];
-            }
+    const getRows = useCallback(() => {
+        if (props.disabled) {
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/components/spreadsheet/table-wrapper.js:301]: fixed

File: src/components/network/use-equipments.js (hunk 1 above)
Comment (archendel) [reply src/components/network/use-equipments.js:13]: fixed

File: src/components/network/use-equipments.js (hunk 2 above)
Comment (archendel) [reply src/components/network/use-equipments.js:16]: fixed

File: src/components/network/use-equipments.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import { useEffect } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { loadEquipments } from 'redux/actions';
+
+export const useEquipments = (equipmentType) => {
+    const dispatch = useDispatch();
```
[the remark this answers is not in the document]
Comment (flomillot) [reply src/components/network/use-equipments.js]: answer : we stay on redux, but maybe switch on createSlice to be able to fetch data with redux and then we wouldn't need that hook ?

**Review by Meklo — APPROVED** [review]

File: src/components/study-container.js [hunk 8]
```diff
@@ -248,11 +252,8 @@ export function StudyContainer({ view, onChangeTab }) {
     // using a ref because this is not used for rendering, it is used in the websocket onMessage()
     const studyParentDirectoriesUuidsRef = useRef([]);
 
-    const network = useSelector((state) => state.network);
     const userName = useSelector((state) => state.user.profile.sub);
     const paramsLoaded = useSelector((state) => state[PARAMS_LOADED]);
-    const [networkLoadingFailMessage, setNetworkLoadingFailMessage] =
```
Comment (flomillot) [inline src/components/study-container.js:254]: You don't manage the loading fail of equipments ?
Now the app is stuck on "loading" if I shut down a service or if we receive an error.

File: src/components/network/use-spreadsheet-equipments.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    useEffect(() => {
+        if (!equipments) {
+            Promise.all(
+                equipmentType.type.fetchers.map((fetcher) =>
+                    fetcher(studyUuid, currentNode.id)
+                )
+            ).then((results) => {
+                const equipments = results.flat();
+                dispatch(loadEquipments(equipmentType.resource, equipments));
+            });
```
Comment (flomillot) [inline src/components/network/use-spreadsheet-equipments.js:24]: No catch ?

**Review by flomillot — COMMENTED** [review]
As discussed, apparently `networkElement` now replace the previous word `equipment`.
@SlimaneAmar should we use this from now for every new dev ?

File: src/redux/reducer.js [hunk 10]
```diff
@@ -904,8 +906,187 @@ export const reducer = createReducer(initialState, {
...
+    }
+    return currentEquipments;
+}
+
+function updateSubstationsAndVoltageLevels(
+    currentSubstations,
+    currentVoltageLevels,
+    newOrUpdatedSubstations
+) {
+    currentSubstations = updateEquipments(
```
Comment (flomillot) [inline src/redux/reducer.js:1051]: You shouldn't create a new const instead and rename it ? Same for VL ?

File: src/components/study-container.js (hunk 8 above)
Comment (archendel) [reply src/components/study-container.js:254]: fixed

File: src/components/network/use-spreadsheet-equipments.js (hunk 9 above)
Comment (archendel) [reply src/components/network/use-spreadsheet-equipments.js]: fixed

File: src/redux/reducer.js (hunk 10 above)
Comment (archendel) [reply src/redux/reducer.js:1052]: fixed

File: src/components/spreadsheet/table-wrapper.js [hunk 11]
```diff
@@ -293,10 +293,19 @@ const TableWrapper = (props) => {
...
     );
 
-    const { equipments, isFetching } = useSpreadsheetEquipments(
+    const { equipments, errorMessage } = useSpreadsheetEquipments(
         TABLES_DEFINITION_INDEXES.get(tabIndex)
     );
 
+    useEffect(() => {
+        if (errorMessage) {
+            snackError({
```
Comment (flomillot) [inline src/components/spreadsheet/table-wrapper.js:302]: To discuss with POs

File: src/redux/reducer.js [hunk 12]
```diff
@@ -1049,11 +1049,13 @@ function updateSubstationsAndVoltageLevels(
     currentVoltageLevels,
     newOrUpdatedSubstations
 ) {
-    currentSubstations = updateEquipments(
+    const updatedSubstations = updateEquipments(
         currentSubstations,
         newOrUpdatedSubstations
     );
 
+    let updatedVoltageLevels = currentVoltageLevels;
```
Comment (flomillot) [inline src/redux/reducer.js:1057]: ?

File: src/redux/reducer.js [hunk 13]
```diff
@@ -1063,13 +1065,13 @@ function updateSubstationsAndVoltageLevels(
             []
         );
 
-        currentVoltageLevels = updateEquipments(
+        updatedVoltageLevels = updateEquipments(
```
Comment (flomillot) [inline src/redux/reducer.js:1068]:
```suggestion
        const updatedVoltageLevels = updateEquipments(
```

File: src/translations/fr.json [hunk 14]
```diff
@@ -1018,5 +1018,6 @@
     "Create": "Créer",
     "ShuntCompensator": "Moyen de compensation statique",
     "AttachingAndSplittingLines": "Piquage et coupure",
-    "GenerationAndLoad": "Production et consommation"
+    "GenerationAndLoad": "Production et consommation",
+    "SpreadsheetFetchError": "Une erreur est survenue lors du chargement des ouvrages"
```
Comment (flomillot) [inline src/translations/fr.json:1022]:
```suggestion
    "SpreadsheetFetchError": "Une erreur est survenue lors du chargement des ouvrages du tableur"
```
Same for english

File: src/components/spreadsheet/table-wrapper.js (hunk 11 above)
Comment (archendel) [reply src/components/spreadsheet/table-wrapper.js:302]: Done, they're OK with this solution

File: src/redux/reducer.js (hunk 12 above)
Comment (archendel) [reply src/redux/reducer.js:1057]: replaced with null

File: src/redux/reducer.js (hunk 13 above)
Comment (archendel) [reply src/redux/reducer.js:1068]: shouldn't be changed

File: src/translations/fr.json (hunk 14 above)
Comment (archendel) [reply src/translations/fr.json:1022]: english : An error occurred while fetching equipments in spreadsheet 
is it ok ?

Comment (flomillot) [reply src/translations/fr.json]: To be the most correct and coherent possible, I would write :

Une erreur est survenue lors du chargement d'ouvrages dans le tableur
An error occurred while fetching equipments in the spreadsheet

But it's fine

**Review by flomillot — APPROVED** [review]

**Pull request merged**